engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java

```diff
@@ -0,0 +1,146 @@
+package org.enso.interpreter.test;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.file.Paths;
+import java.util.Map;
+import org.enso.polyglot.RuntimeOptions;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Language;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PolyglotErrorTest {
+  private Context ctx;
+  private Value panic;
+
+  public static String bar(Object o) {
+    return "[[" + o + "]]";
+  }
+
+  @Before
+  public void prepareCtx() throws Exception {
+    this.ctx = Context.newBuilder()
+      .allowExperimentalOptions(true)
+      .allowIO(true)
+      .allowAllAccess(true)
+      .logHandler(new ByteArrayOutputStream())
+      .option(
+        RuntimeOptions.LANGUAGE_HOME_OVERRIDE,
+        Paths.get("../../distribution/component").toFile().getAbsolutePath()
+      ).build();
+    final Map<String, Language> langs = ctx.getEngine().getLanguages();
+    assertNotNull("Enso found: " + langs, langs.get("enso"));
+
+    var code = """
```

Comment:
Sorry that I'm writing post-mortem but I've only now figured out one more way to break this:

```
from Standard.Base import all
polyglot java import bar.Foo


type Ca_Type
    Ca x

    to_text self = Panic.throw "PANIK"

type Cd_Type
    Cd x

    to_text self = 23

type Cf_Type
    Cf x

    to_text self = self

main =
    z = Cd_Type.Cd 42
    Foo.bar z

    y = Ca_Type.Ca 42
    Foo.bar y

    x = Cf_Type.Cf 42
    Foo.bar x
```

on current develop (after this PR), yields:
```
[|Error in method `to_text` of [Cd 42]: Expected Text but got 23|]
[|Panic in method `to_text` of [Ca 42]: 'PANIK'|]
[|Error in toString()|]
```

I was testing this because I had a hunch that `to_text self = self` could lead to an infinite loop with the new logic. Apparently it doesn't, it just errors. For some reason though the error is different than what we handle so it still fallsback to the default Truffle `Error in toString()`.

I don't think it's a bad thing because it is a very rare case, I do not know how to repro this other than by returning `self`.

Still I thought it's nice to at least keep note of this fact.

Reply:
@radeusgd, please report the missing test case in pivotal as a bug.

Reply:
OK

https://www.pivotaltracker.com/story/show/184324080